test: add test-domain-exit-dispose-again back

misterdjules opened the pull request:
1c8584997346d549dd5ff4bb787f48f52440a9cb "fixed"
test-domain-exit-dispose-again by changing its logic to test that
process.domain was cleared properly in case an error was thrown from a
timer's callback.

However, it became clear when reviewing a recent change that refactors
lib/timers.js that it was not quite the intention of the original test.
Thus, this change adds the original implementation of
test-domain-exit-dispose-again back, with comments that make its
implementation easier to understand.

It also preserve the changes made by
1c8584997346d549dd5ff4bb787f48f52440a9cb, but it moves them to a new
test file named test-timers-reset-process-domain-on-throw.js.

/cc @Fishrock123 This is the test I mentioned in https://github.com/nodejs/node/pull/4007#discussion_r46505745.

misterdjules commented:
CI tests [running](https://ci.nodejs.org/job/node-test-pull-request/985/).

Fishrock123 commented:
Tests seem fine, I'll test it on the PR with and without the related code when I get the chance.

Fishrock123 commented:
Note: my timers branch does not have https://github.com/nodejs/node/commit/1c8584997346d549dd5ff4bb787f48f52440a9cb, and the old version doesn't catch the relevant change in timers.

Fishrock123 commented on test/parallel/test-timers-reset-process-domain-on-throw.js:
```diff
@@ -0,0 +1,45 @@
+  if (process.domain !== null) {
+    console.log('process.domain should be null, but instead is:',
+      process.domain);
+    process.exit(1);
```
Usually we avoid using `process.exit()` in tests because it is rather immediate, is there a way this could be switched to Throw?

misterdjules replied:
The problem is explained by the comment above:

```
// Do not use assert here, as it throws errors and if a domain with an error
// handler is active, then asserting wouldn't make the test fail.
```

maybe this comment should be move right above the call to `process.exit(1)`?

Fishrock123 replied:
Oh, probably. And of course, domains, oopsie.

misterdjules replied:
Sounds good, I'll move it closer to the call to `process.exit`.

Fishrock123 commented on test/parallel/test-timers-reset-process-domain-on-throw.js:
```diff
@@ -0,0 +1,45 @@
+  // Do not use assert here, as it throws errors and if a domain with an error
+  // handler is active, then asserting wouldn't make the test fail.
+  if (process.domain !== null) {
+    console.log('process.domain should be null, but instead is:',
```
Perhaps it would be more clear as `'process.domain should be null on timer'`?

misterdjules replied:
What do you mean by "on timer"? `process.domain` is a process-wide variable, so specifying "on timer" doesn't necessarily make sense.

Fishrock123 replied:
"inside timer callback"?

misterdjules replied:
It should be null in _that specific_ timer's callback, not in _every_ timer's callbacks. I can rephrase this as `process.domain should be null in this timer callabck, but instead id:` if that helps.

misterdjules commented:
@Fishrock123 Just to make sure we're on the same page: do you mean that the version of `test/parallel/test-domain-exit-dispose-again.js` prior to https://github.com/nodejs/node/commit/1c8584997346d549dd5ff4bb787f48f52440a9cb doesn't fail if `process.domain` is not [reset after a timer's callback threw](https://github.com/nodejs/node/blob/master/lib/timers.js#L102)? That's what I'd expect, and I would expect both https://github.com/nodejs/node/commit/1c8584997346d549dd5ff4bb787f48f52440a9cb and the changes in this PR to make that case fail.

misterdjules commented:
/ping @Fishrock123 Did you have some time to look at my answers to your questions/comments?

misterdjules commented:
@Fishrock123 Pushed a new commit that I believe addresses your review comments. If that looks good to you, I'll squash the two commits in one. Thanks!

Fishrock123 commented:
> @Fishrock123 Just to make sure we're on the same page: do you mean that the version of `test/parallel/test-domain-exit-dispose-again.js` prior to https://github.com/nodejs/node/commit/1c8584997346d549dd5ff4bb787f48f52440a9cb doesn't fail if `process.domain` is not [reset after a timer's callback threw](https://github.com/nodejs/node/blob/master/lib/timers.js#L102)? That's what I'd expect, and I would expect both https://github.com/nodejs/node/commit/1c8584997346d549dd5ff4bb787f48f52440a9cb and the changes in this PR to make that case fail.

I don't understand the details, but ok!

Rubber-stamp LGTM as I don't feel like I can adequately assess the tests' details, but since this is just re-introducing an old test it should be fine.

CI again, for good measure: https://ci.nodejs.org/job/node-test-pull-request/1144/

misterdjules commented:
Thank you @Fishrock123. Landed in 1453b872aea1dc89b9d1fe1c8dabe392ec1e9e89.
